run.go

```diff
@@ -139,38 +135,102 @@ var commandInspect = cli.Command{
 	Action: actionInspect,
 }
 
-func guessType(filename string) string {
-	switch {
-	case strings.Contains(filename, "://"):
-		return TypeURL
-	case strings.HasSuffix(filename, ".js"):
-		return TypeJS
-	default:
-		return ""
+func looksLikeURL(str []byte) bool {
+	// Try to parse string as URL
+	u, err := url.Parse(strings.TrimSpace(string(str)))
+	if err != nil || u.Scheme == "" {
+		return false
 	}
+	return true
 }
 
-func makeRunner(filename, t string) (lib.Runner, error) {
-	if t == TypeAuto {
-		t = guessType(filename)
+func getSrcData(arg, t string) (*lib.SourceData, error) {
+	srcdata := []byte("")
+	srctype := t
+	filename := arg
+	var err error
+	const cmdline = "[cmdline]"
+	// special case name “-“ will always cause src data to be read from file STDIN
+	if arg == "-" {
+		srcdata, err = ioutil.ReadAll(os.Stdin)
+		if err != nil {
+			return nil, err
+		}
+	} else {
+		// Deduce how to get src data
+		switch srctype {
+		case lib.TypeAuto:
```

Comment:
You could keep this significantly DRYer by just checking if it's TypeAuto, and if it is, just sniffing out what the actual type is. Keep the switch for the concrete types.